File: src/main/java/com/redbox/domain/attach/service/AttachFileService.java
package com.redbox.domain.attach.service;

import com.redbox.domain.attach.dto.AttachFileResponse;
import com.redbox.domain.attach.entity.AttachFile;
import com.redbox.domain.attach.entity.Category;
import com.redbox.domain.attach.exception.AttachFileNotFoundException;
import com.redbox.domain.attach.exception.FileNotBelongException;
import com.redbox.domain.attach.repository.AttachFileRepository;
import com.redbox.domain.attach.strategy.FileAttachStrategy;
import com.redbox.domain.attach.strategy.FileAttachStrategyFactory;
import com.redbox.domain.notice.repository.NoticeRepository;
import com.redbox.global.infra.s3.S3Service;
import com.redbox.global.util.FileUtils;
import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

@Service
@RequiredArgsConstructor
public class AttachFileService {

    private final AttachFileRepository attachFileRepository;
    private final S3Service s3Service;
    private final NoticeRepository noticeRepository;
    private final FileAttachStrategyFactory fileAttachStrategyFactory;
    private final RedisTemplate<String, Object> redisTemplate;
    private static final String NOTICE_DETAIL_KEY = "notices:detail:%d";    // 공지사항 상세글

    public String getFileDownloadUrl(Long postId, Long fileId) {
        AttachFile attachFile = attachFileRepository.findById(fileId)
                .orElseThrow(AttachFileNotFoundException::new);

        validateFileOwnership(attachFile, postId);

        // PreSignedURL 생성
        return s3Service.generatePresignedUrl(
                attachFile.getCategory(),
                postId,
                attachFile.getNewFilename(),
                attachFile.getOriginalFilename()
        );
    }

    private void validateFileOwnership(AttachFile attachFile, Long postId) {
        if (!attachFile.belongToPost(postId)) {
            throw new FileNotBelongException();
        }
    }

    @Transactional
    public AttachFileResponse addFile(Category category, Long postId, MultipartFile file) {
        // S3에 파일 업로드
        String newFilename = FileUtils.generateNewFilename();
        String extension = FileUtils.getExtension(file);
        String fullFilename = newFilename + "." + extension;
        s3Service.uploadFile(file, category, postId, fullFilename);

        // 카테고리에 맞는 전략 사용
        FileAttachStrategy strategy = fileAttachStrategyFactory.getStrategy(category);
        AttachFile attachFile = strategy.attach(postId, file.getOriginalFilename(), fullFilename);
        attachFileRepository.save(attachFile);

        if (category.equals(Category.NOTICE)) {
            redisTemplate.delete(String.format(NOTICE_DETAIL_KEY, postId));
        }
        return new AttachFileResponse(attachFile);
    }

    @Transactional
    public void removeFile(Category category, Long postId, Long fileId) {
        AttachFile attachFile = attachFileRepository.findById(fileId)
                .orElseThrow(AttachFileNotFoundException::new);

        validateFileOwnership(attachFile, postId);

        s3Service.deleteFile(category, postId, attachFile.getNewFilename());

        if (category.equals(Category.NOTICE)) {
            redisTemplate.delete(String.format(NOTICE_DETAIL_KEY, postId));
        }
        attachFileRepository.delete(attachFile);
    }
}

File: migration/src/main/kotlin/com/redbox/domain/community/attach/service/AttachFileService.kt
package com.redbox.domain.community.attach.service

import com.redbox.domain.community.attach.dto.AttachFileResponse
import com.redbox.domain.community.attach.entity.Category
import com.redbox.domain.community.attach.repository.AttachFileRepository
import com.redbox.domain.community.attach.strategy.FileAttachStrategy
import com.redbox.domain.community.attach.strategy.FileAttachStrategyFactory
import com.redbox.global.infra.s3.S3Service
import com.redbox.global.util.FileUtils
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile

@Service
class AttachFileService(
    private val s3Service: S3Service,
    private val fileAttachStrategyFactory: FileAttachStrategyFactory,
    private val attachFileRepository: AttachFileRepository,
) {
    @Transactional
    fun addFile(category: Category, postId: Long?, file: MultipartFile): AttachFileResponse {
        // S3에 파일 업로드
        val newFilename = FileUtils.generateNewFilename()
        val extension = FileUtils.getExtension(file)
        val fullFilename = "$newFilename.$extension"
        postId?.let { s3Service.uploadFile(file, category, it, fullFilename) }

        // 카테고리에 맞는 전략 사용
        val strategy: FileAttachStrategy? = fileAttachStrategyFactory.getStrategy(category)
        val attachFile = strategy?.attach(postId, file.originalFilename, fullFilename)
            ?: throw IllegalArgumentException("No strategy found for category: $category")
        attachFileRepository.save(attachFile)

        /*if (category.equals(Category.NOTICE)) {
            redisTemplate.delete(String.format(AttachFileService.NOTICE_DETAIL_KEY, postId))
        }*/

        return AttachFileResponse(attachFile)
    }
}